api/envoy/config/rbac/v2alpha/rbac.proto

```diff
@@ -2,42 +2,49 @@ syntax = "proto3";
 
 import "validate/validate.proto";
 import "envoy/api/v2/core/address.proto";
-import "envoy/type/string_match.proto";
+import "envoy/api/v2/route/route.proto";
 
 package envoy.config.rbac.v2alpha;
 option go_package = "v2alpha";
 
-// Role Based Access Control (RBAC) provides service-level and method-level access control for a service.
-// The RBAC engine authorizes a request by evaluating the request context (expressed in the form of
-// :ref: `AttributeContext <envoy_api_msg_service.auth.v2alpha.AttributeContext>`) against the RBAC policies.
+// Role Based Access Control (RBAC) provides service-level and method-level access control for a
+// service. The RBAC engine authorizes a request by evaluating the request context (a combination of
+// the connection data and HTTP request headers) against the RBAC policies.
 //
-// RBAC policies are additive. The policies are examined in order. A request is allowed once a matching policy
-// is found (suppose the `action` is ALLOW).
+// RBAC policies are additive. The policies are examined in order. If the `action` is ALLOW, a
+// request is allowed once a policy is matched. If the `action` is DENY, a request is disallowed
+// once a policy is matched.
 //
 // Here is an example of RBAC configuration. It has two policies:
-// * Service account "cluster.local/ns/default/sa/admin" has full access (empty permission entry means full access)
-//   to the service.
-// * Any user (empty principal entry means any user) can read ("GET") the service at paths with prefix "/products" or
-//   suffix "/reviews" when request header "version" set to either "v1" or "v2".
+//
+// * Service account "cluster.local/ns/default/sa/admin" or "cluster.local/ns/default/sa/superuser"
+//   has full access to take any action.
+//
+// * Any user can read ("GET") from the service at paths with prefix "/products" when the
+//   destination port is either 80 or 443.
+//
+// .. code-block:: yaml
 //
 //   action: ALLOW
 //   policies:
 //     "service-admin":
 //       permissions:
-//       -
+//         - any: true
 //       principals:
-//         authenticated:
-//           name: "cluster.local/ns/default/sa/admin"
+//         - authenticated: { name: "cluster.local/ns/default/sa/admin" }
+//         - authenticated: { name: "cluster.local/ns/default/sa/superuser" }
 //     "product-viewer":
 //       permissions:
-//       - paths: [prefix: "/products", suffix: "/reviews"]
-//         methods: ["GET"]
-//         conditions:
-//         - header:
-//             key: "version"
-//             values: [simple: "v1", simple: "v2"]
+//           - and_rules:
+//               rules:
+//                 - header: { name: ":method", exact_match: "GET" }
+//                 - header: { name: ":path", regex_match: "/products(/.*)?" }
+//                 - or_rules:
+//                     rules:
+//                       - destination_port: 80
+//                       - destination_port: 443
```

Comment:
An RBAC role is a set of permissions. Each permission defines an allowed action. So by definition, it is "OR" relationship between permissions. And within a single permission, the rules are "AND"ed together. Hence, 
1. We do not need top-level "or_rules" in a permission, because if you define "A OR B" within a permission, it is better to create two permissions "A" and "B".
2. We do not have to specify "and_rules" explicitly because within a permission, the rules are always ANDed. 
3. I see that you have changed the semantics of "attribute-->a list of allowed values" to specific or_rules (attribute==value1 or attribute==value2). So the example you showed (removed top-level "and_rules" based on the above comment).

```
permissions:
- header: { name: ":method", exact_match: "GET" }
  header: { name: ":path", regex_match: "/products(/.*)?" }
  or_rules:
    rules:
      - destination_port: 80
      - destination_port: 443
```
vs. attribute->a list of allowed values model.
```
permissions:
- header: { name: ":method", exact_match: "GET" }
  header: { name: ":path", regex_match: "/products(/.*)?" }
  destination_port: [80, 443]
```
Regarding 3, just from user experience perspective, I feel that the second (attribute->a list of allowed value) presentation is simpler. What do you think? But if it creates difficulty/confusion in implementation, I am fine either way.  

Same comments apply to "principals". By definition, it is "OR" relationship between different principals. Within a single "principal", the rules are "AND"ed together. Hence, the three points I mentioned above apply to "principals" fields too.

Reply:
I believe it's been discussed before, but the ergonomics of the configs here are for the Envoy's implementation itself, not so much the user. User-friendlier structures can be easily morphed to this model.

As for why there is support for AND/OR similar to the original implementation of the auth protos, it's to make it very explicit of what the intent is of how the rules within a permission/principal apply. While I agree from a user's perspective that the top-level `and_rules` are redundant, it makes it clear how each sub-rule is evaluated and leaves nothing up to convention. The existence of the and/or_rules is not mutually exclusive with your approach of splitting out each individual permission/principal within a policy, but it does have the potential to eliminate an explosion of rules which could be more succinctly described with composed ANDs/ORs. 

This form, IMO, strikes a balance between the original intent as you proposed, plus makes it easier and clearer how it should be interpreted by the engine, as well as provides flexibility for use cases that go beyond a typical RBAC design.

Reply:
Thanks for the explanation. Per our offline discussion, I am fine with the proto changes.